package soa;

message Control {
    optional string token = 1;
}

message ActionRequest {
    required string action = 1;
    optional ActionRequestParams params = 2;
}

message ActionRequestParams {

    // give us room to extend the ActionRequestParams message
    extensions 100 to max;
}

message ActionResponseResult {
    optional bool success = 1;
    repeated string errors = 2;
    repeated ErrorDetail error_details = 3;

    message ErrorDetail {
        optional string error_key = 1;
        optional string error_detail = 2;
    }

    // give us room to extend the ActionResponseResult
    extensions 100 to max;
}

message ActionResponse {
    required string action = 1;
    optional ActionResponseResult response = 2;
}

message ServiceRequest {
    required Control control = 1;
    repeated ActionRequest actions = 2;
}

message ServiceResponse {
    required Control control = 1;
    repeated ActionResponse actions = 2;
}
